File: app/src/main/java/wangdaye/com/geometricweather/weather/services/MfWeatherService.java
package wangdaye.com.geometricweather.weather.services;

import android.content.Context;
import android.text.TextUtils;

import androidx.annotation.NonNull;

import java.io.IOException;
import java.util.ArrayList;
import java.util.List;

import javax.inject.Inject;

import io.reactivex.Observable;
import io.reactivex.disposables.CompositeDisposable;
import wangdaye.com.geometricweather.BuildConfig;
import wangdaye.com.geometricweather.common.basic.models.Location;
import wangdaye.com.geometricweather.common.rxjava.BaseObserver;
import wangdaye.com.geometricweather.common.rxjava.ObserverContainer;
import wangdaye.com.geometricweather.common.rxjava.SchedulerTransformer;
import wangdaye.com.geometricweather.settings.ConfigStore;
import wangdaye.com.geometricweather.settings.SettingsManager;
import wangdaye.com.geometricweather.weather.apis.AtmoAuraIqaApi;
import wangdaye.com.geometricweather.weather.apis.MfWeatherApi;
import wangdaye.com.geometricweather.weather.converters.MfResultConverter;
import wangdaye.com.geometricweather.weather.json.atmoaura.AtmoAuraQAResult;
import wangdaye.com.geometricweather.weather.json.mf.MfCurrentResult;
import wangdaye.com.geometricweather.weather.json.mf.MfEphemerisResult;
import wangdaye.com.geometricweather.weather.json.mf.MfForecastResult;
import wangdaye.com.geometricweather.weather.json.mf.MfForecastV2Result;
import wangdaye.com.geometricweather.weather.json.mf.MfLocationResult;
import wangdaye.com.geometricweather.weather.json.mf.MfRainResult;
import wangdaye.com.geometricweather.weather.json.mf.MfWarningsResult;

/**
 * Mf weather service.
 */

public class MfWeatherService extends WeatherService {

    private final MfWeatherApi mMfApi;
    private final AtmoAuraIqaApi mAtmoAuraApi;
    private final CompositeDisposable mCompositeDisposable;

    private static final String CONFIG_NAME_LOCAL = "LOCAL_PREFERENCE_MF";
    private static final String KEY_OLD_DISTRICT = "OLD_DISTRICT";
    private static final String KEY_OLD_CITY = "OLD_CITY";
    private static final String KEY_OLD_PROVINCE = "OLD_PROVINCE";
    private static final String KEY_OLD_KEY = "OLD_KEY";

    private static class CacheLocationRequestCallback implements RequestLocationCallback {

        private final Context mContext;
        private final @NonNull RequestLocationCallback mCallback;

        CacheLocationRequestCallback(Context context, @NonNull RequestLocationCallback callback) {
            mContext = context;
            mCallback = callback;
        }

        @Override
        public void requestLocationSuccess(String query, List<Location> locationList) {
            if (!TextUtils.isEmpty(locationList.get(0).getCityId())) {
                ConfigStore.getInstance(mContext, CONFIG_NAME_LOCAL)
                        .edit()
                        .putString(KEY_OLD_KEY, locationList.get(0).getCityId())
                        .apply();
            }
            mCallback.requestLocationSuccess(query, locationList);
        }

        @Override
        public void requestLocationFailed(String query) {
            ConfigStore.getInstance(mContext, CONFIG_NAME_LOCAL)
                    .edit()
                    .putString(KEY_OLD_DISTRICT, "")
                    .putString(KEY_OLD_CITY, "")
                    .putString(KEY_OLD_PROVINCE, "")
                    .putString(KEY_OLD_KEY, "")
                    .apply();
            mCallback.requestLocationFailed(query);
        }
    }

    private static class EmptyAtmoAuraQAResult extends AtmoAuraQAResult {
    }

    private static class EmptyWarningsResult extends MfWarningsResult {
    }

    @Inject
    public MfWeatherService(MfWeatherApi mfApi, AtmoAuraIqaApi atmoApi,
                            CompositeDisposable disposable) {
        mMfApi = mfApi;
        mAtmoAuraApi = atmoApi;
        mCompositeDisposable = disposable;
    }

    @Override
    public void requestWeather(Context context, Location location, @NonNull RequestWeatherCallback callback) {
        String languageCode = SettingsManager.getInstance(context).getLanguage().getCode();

        Observable<MfCurrentResult> current = mMfApi.getCurrent(
                location.getLatitude(), location.getLongitude(), languageCode, BuildConfig.MF_WSFT_KEY);

        Observable<MfForecastResult> forecast = mMfApi.getForecast(
                location.getLatitude(), location.getLongitude(), languageCode, BuildConfig.MF_WSFT_KEY);

        // TODO: Will allow us to display forecast for day and night in daily
        //Observable<MfForecastResult> dayNightForecast = api.getForecastInstants(
        //        location.getLatitude(), location.getLongitude(), languageCode, "afternoon,night", BuildConfig.MF_WSFT_KEY);

        Observable<MfEphemerisResult> ephemeris = mMfApi.getEphemeris(
                location.getLatitude(), location.getLongitude(), "en", BuildConfig.MF_WSFT_KEY);
        // English required to convert moon phase

        Observable<MfRainResult> rain = mMfApi.getRain(
                location.getLatitude(), location.getLongitude(), languageCode, BuildConfig.MF_WSFT_KEY);

        Observable<MfWarningsResult> warnings = mMfApi.getWarnings(
                location.getProvince(), null, BuildConfig.MF_WSFT_KEY
        ).onExceptionResumeNext(
                // FIXME: Will not report warnings if current location was searched through AccuWeather search because "province" is not the department
                Observable.create(emitter -> emitter.onNext(new EmptyWarningsResult()))
        );

        Observable<AtmoAuraQAResult> aqiAtmoAura;
        if (location.getProvince() == null) {
            aqiAtmoAura = Observable.create(emitter -> emitter.onNext(new EmptyAtmoAuraQAResult()));
        } else if (location.getProvince().equals("Auvergne-Rhône-Alpes") || location.getProvince().equals("01")
                || location.getProvince().equals("03") || location.getProvince().equals("07")
                || location.getProvince().equals("15") || location.getProvince().equals("26")
                || location.getProvince().equals("38") || location.getProvince().equals("42")
                || location.getProvince().equals("43") || location.getProvince().equals("63")
                || location.getProvince().equals("69") || location.getProvince().equals("73")
                || location.getProvince().equals("74")) {
            aqiAtmoAura = mAtmoAuraApi.getQAFull(
                    BuildConfig.IQA_ATMO_AURA_KEY,
                    String.valueOf(location.getLatitude()),
                    String.valueOf(location.getLongitude())
            ).onExceptionResumeNext(
                    Observable.create(emitter -> emitter.onNext(new EmptyAtmoAuraQAResult()))
            );
        } else {
            aqiAtmoAura = Observable.create(emitter -> emitter.onNext(new EmptyAtmoAuraQAResult()));
        }

        Observable.zip(current, forecast, ephemeris, rain, warnings, aqiAtmoAura,
                (mfCurrentResult, mfForecastResult, mfEphemerisResult, mfRainResult, mfWarningResults, aqiAtmoAuraResult) -> MfResultConverter.convert(
                        context,
                        location,
                        mfCurrentResult,
                        mfForecastResult,
                        mfEphemerisResult,
                        mfRainResult,
                        mfWarningResults,
                        aqiAtmoAuraResult instanceof EmptyAtmoAuraQAResult ? null : aqiAtmoAuraResult
                )
        ).compose(SchedulerTransformer.create())
                .subscribe(new ObserverContainer<>(mCompositeDisposable, new BaseObserver<WeatherResultWrapper>() {
                    @Override
                    public void onSucceed(WeatherResultWrapper wrapper) {
                        if (wrapper.result != null) {
                            location.setWeather(wrapper.result);
                            callback.requestWeatherSuccess(location);
                        } else {
                            onFailed();
                        }
                    }

                    @Override
                    public void onFailed() {
                        callback.requestWeatherFailed(location);
                    }
                }));
    }

    @Override
    @NonNull
    public List<Location> requestLocation(Context context, String query) {
        List<MfLocationResult> resultList = null;
        try {
            resultList = mMfApi.callWeatherLocation(query, 48.86d, 2.34d, BuildConfig.MF_WSFT_KEY).execute().body();
        } catch (IOException e) {
            e.printStackTrace();
        }

        List<Location> locationList = new ArrayList<>();
        if (resultList != null && resultList.size() != 0) {
            for (MfLocationResult r : resultList) {
                if (r.postCode != null) {
                    locationList.add(MfResultConverter.convert(null, r));
                }
            }
        }
        return locationList;
    }

    @Override
    public void requestLocation(Context context, Location location,
                                @NonNull RequestLocationCallback callback) {
        ConfigStore config = ConfigStore.getInstance(context, CONFIG_NAME_LOCAL);
        String oldDistrict = config.getString(KEY_OLD_DISTRICT, "");
        String oldCity = config.getString(KEY_OLD_CITY, "");
        String oldProvince = config.getString(KEY_OLD_PROVINCE, "");
        String oldKey = config.getString(KEY_OLD_KEY, "");

        if (location.hasGeocodeInformation()
                && queryEqualsIgnoreEmpty(location.getDistrict(), oldDistrict)
                && queryEquals(location.getCity(), oldCity)
                && queryEquals(location.getProvince(), oldProvince)
                && queryEquals(location.getCityId(), oldKey)) {
            List<Location> locationList = new ArrayList<>();
            locationList.add(location);
            callback.requestLocationSuccess(
                    location.getLatitude() + "," + location.getLongitude(),
                    locationList
            );
            return;
        }

        config.edit()
                .putString(KEY_OLD_DISTRICT, location.getDistrict())
                .putString(KEY_OLD_CITY, location.getCity())
                .putString(KEY_OLD_PROVINCE, location.getProvince())
                .apply();

        String languageCode = SettingsManager.getInstance(context).getLanguage().getCode();
        final CacheLocationRequestCallback finalCallback = new CacheLocationRequestCallback(context, callback);

        mMfApi.getForecastV2(
                location.getLatitude(),
                location.getLongitude(),
                languageCode,
                BuildConfig.MF_WSFT_KEY
        ).compose(SchedulerTransformer.create())
                .subscribe(new ObserverContainer<>(mCompositeDisposable, new BaseObserver<MfForecastV2Result>() {
                    @Override
                    public void onSucceed(MfForecastV2Result mfForecastV2Result) {
                        if (mfForecastV2Result != null) {
                            List<Location> locationList = new ArrayList<>();
                            if (mfForecastV2Result.properties.insee != null) {
                                locationList.add(MfResultConverter.convert(null, mfForecastV2Result));
                            }
                            // FIXME: Caching geo position
                            finalCallback.requestLocationSuccess(
                                    location.getLatitude() + "," + location.getLongitude(), locationList);
                        } else {
                            onFailed();
                        }
                    }

                    @Override
                    public void onFailed() {
                        // FIXME: Caching geo position
                        finalCallback.requestLocationFailed(
                                location.getLatitude() + "," + location.getLongitude());
                    }
                }));
    }

    public void requestLocation(Context context, String query,
                                @NonNull RequestLocationCallback callback) {
        mMfApi.getWeatherLocation(query, 48.86d, 2.34d, BuildConfig.MF_WSFT_KEY)
                .compose(SchedulerTransformer.create())
                .subscribe(new ObserverContainer<>(mCompositeDisposable, new BaseObserver<List<MfLocationResult>>() {
                    @Override
                    public void onSucceed(List<MfLocationResult> mfLocationResults) {
                        if (mfLocationResults != null && mfLocationResults.size() != 0) {
                            List<Location> locationList = new ArrayList<>();
                            for (MfLocationResult r : mfLocationResults) {
                                if (r.postCode != null) {
                                    locationList.add(MfResultConverter.convert(null, r));
                                }
                            }
                            callback.requestLocationSuccess(query, locationList);
                        } else {
                            callback.requestLocationFailed(query);
                        }
                    }

                    @Override
                    public void onFailed() {
                        callback.requestLocationFailed(query);
                    }
                }));
    }

    @Override
    public void cancel() {
        mCompositeDisposable.clear();
    }

    private boolean queryEquals(String a, String b) {
        if (!TextUtils.isEmpty(a) && !TextUtils.isEmpty(b)) {
            return a.equals(b);
        }
        return false;
    }

    private boolean queryEqualsIgnoreEmpty(String a, String b) {
        if (TextUtils.isEmpty(a) && TextUtils.isEmpty(b)) {
            return true;
        }
        if (!TextUtils.isEmpty(a) && !TextUtils.isEmpty(b)) {
            return a.equals(b);
        }
        return false;
    }
}

File: app/src/main/java/wangdaye/com/geometricweather/weather2/services/MfWeatherService.kt
package wangdaye.com.geometricweather.weather2.services

import android.content.Context
import android.text.TextUtils
import dagger.hilt.android.qualifiers.ApplicationContext
import kotlinx.coroutines.async
import kotlinx.coroutines.coroutineScope
import wangdaye.com.geometricweather.BuildConfig
import wangdaye.com.geometricweather.common.basic.models.Location
import wangdaye.com.geometricweather.common.basic.models.weather.Weather
import wangdaye.com.geometricweather.settings.ConfigStore
import wangdaye.com.geometricweather.settings.SettingsManager
import wangdaye.com.geometricweather.weather2.apis.AtmoAuraIqaApi
import wangdaye.com.geometricweather.weather2.apis.MfWeatherApi
import wangdaye.com.geometricweather.weather2.converters.MfResultConverter
import wangdaye.com.geometricweather.weather2.json.mf.*
import java.util.*
import javax.inject.Inject

/**
 * Mf weather service.
 */
class MfWeatherService @Inject constructor(@ApplicationContext context: Context,
                                           private val mfApi: MfWeatherApi,
                                           private val atmoAuraApi: AtmoAuraIqaApi) : WeatherService() {

    companion object {
        private const val CONFIG_NAME_LOCAL = "LOCAL_PREFERENCE_MF"
        private const val KEY_OLD_DISTRICT = "OLD_DISTRICT"
        private const val KEY_OLD_CITY = "OLD_CITY"
        private const val KEY_OLD_PROVINCE = "OLD_PROVINCE"
        private const val KEY_OLD_KEY = "OLD_KEY"
    }

    private val config = ConfigStore.getInstance(context, CONFIG_NAME_LOCAL)

    override suspend fun getWeather(context: Context, location: Location): Weather? = coroutineScope {
        val languageCode = SettingsManager.getInstance(context).getLanguage().code

        val current = async {
            mfApi.getCurrent(
                    location.latitude.toDouble(),
                    location.longitude.toDouble(),
                    languageCode,
                    BuildConfig.MF_WSFT_KEY
            )
        }
        val forecast = async {
            mfApi.getForecast(
                    location.latitude.toDouble(),
                    location.longitude.toDouble(),
                    languageCode,
                    BuildConfig.MF_WSFT_KEY
            )
        }

        // TODO: Will allow us to display forecast for day and night in daily
        //Observable<MfForecastResult> dayNightForecast = api.getForecastInstants(
        //        location.getLatitude(), location.getLongitude(), languageCode, "afternoon,night", BuildConfig.MF_WSFT_KEY);
        val ephemeris = async {
            mfApi.getEphemeris(
                    location.latitude.toDouble(),
                    location.longitude.toDouble(),
                    "en",
                    BuildConfig.MF_WSFT_KEY
            )
        }
        // English required to convert moon phase
        val rain = async {
            mfApi.getRain(
                    location.latitude.toDouble(),
                    location.longitude.toDouble(),
                    languageCode,
                    BuildConfig.MF_WSFT_KEY
            )
        }
        val warnings = async {
            mfApi.getWarnings(location.province, null, BuildConfig.MF_WSFT_KEY)
        }
        val aqiAtmoAura = async {
            try {
                if (location.province == "Auvergne-Rhône-Alpes"
                        || location.province == "01"
                        || location.province == "03"
                        || location.province == "07"
                        || location.province == "15"
                        || location.province == "26"
                        || location.province == "38"
                        || location.province == "42"
                        || location.province == "43"
                        || location.province == "63"
                        || location.province == "69"
                        || location.province == "73"
                        || location.province == "74") {
                    atmoAuraApi.getQAFull(
                            BuildConfig.IQA_ATMO_AURA_KEY,
                            location.latitude.toString(),
                            location.longitude.toString()
                    )
                }
            } catch (e: Exception) {
                e.printStackTrace()
            }
            null
        }

        return@coroutineScope MfResultConverter.convert(
                context,
                location,
                current.await(),
                forecast.await(),
                ephemeris.await(),
                rain.await(),
                warnings.await(),
                aqiAtmoAura.await()
        ).result
    }

    override suspend fun getLocation(context: Context, query: String): List<Location> = coroutineScope {
        val resultList = mfApi.callWeatherLocation(query, 48.86, 2.34, BuildConfig.MF_WSFT_KEY)

        val locationList: MutableList<Location> = ArrayList()
        for (r in resultList) {
            if (r.postCode != null) {
                locationList.add(MfResultConverter.convert(null, r))
            }
        }
        return@coroutineScope locationList
    }

    override suspend fun getLocation(context: Context, location: Location): List<Location> = coroutineScope {
        val oldDistrict = config.getString(KEY_OLD_DISTRICT, "")
        val oldCity = config.getString(KEY_OLD_CITY, "")
        val oldProvince = config.getString(KEY_OLD_PROVINCE, "")
        val oldKey = config.getString(KEY_OLD_KEY, "")

        if (location.hasGeocodeInformation()
                && queryEqualsIgnoreEmpty(location.district, oldDistrict)
                && queryEquals(location.city, oldCity)
                && queryEquals(location.province, oldProvince)
                && queryEquals(location.cityId, oldKey)) {

            return@coroutineScope ArrayList<Location>()
        }
        config.edit()
                .putString(KEY_OLD_DISTRICT, location.district)
                .putString(KEY_OLD_CITY, location.city)
                .putString(KEY_OLD_PROVINCE, location.province)
                .apply()

        val languageCode = SettingsManager.getInstance(context).getLanguage().code

        val result = mfApi.getForecastV2(
                location.latitude.toDouble(),
                location.longitude.toDouble(),
                languageCode,
                BuildConfig.MF_WSFT_KEY
        )

        val locationList = ArrayList<Location>()
        if (result.properties.insee != null) {
            locationList.add(MfResultConverter.convert(null, result))
        }

        if (locationList.isEmpty()) {
            config.edit()
                    .putString(KEY_OLD_DISTRICT, "")
                    .putString(KEY_OLD_CITY, "")
                    .putString(KEY_OLD_PROVINCE, "")
                    .putString(KEY_OLD_KEY, "")
                    .apply()
        } else if (!TextUtils.isEmpty(locationList[0].cityId)) {
            config.edit()
                    .putString(KEY_OLD_KEY, locationList[0].cityId)
                    .apply()
        }

        return@coroutineScope locationList
    }

    private fun queryEquals(a: String?, b: String?): Boolean {
        return if (!TextUtils.isEmpty(a) && !TextUtils.isEmpty(b)) {
            a == b
        } else {
            false
        }
    }

    private fun queryEqualsIgnoreEmpty(a: String?, b: String?): Boolean {
        if (TextUtils.isEmpty(a) && TextUtils.isEmpty(b)) {
            return true
        }
        return queryEquals(a, b)
    }
}